File: core/src/main/kotlin/okuri/core/controller/block/BlockTypeController.kt
package okuri.core.controller.block

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import okuri.core.models.block.BlockType
import okuri.core.models.block.request.CreateBlockTypeRequest
import okuri.core.service.block.BlockTypeService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/v1/block/schema")
@Tag(name = "Block Type Management", description = "Endpoints for managing block types and their configurations")
class BlockTypeController(
    private val blockTypeService: BlockTypeService
) {

    @PostMapping("/")
    @Operation(
        summary = "Create a new block type",
        description = "Creates and publishes a new block type based on the provided request data."
    )
    @ApiResponses(
        ApiResponse(responseCode = "201", description = "Block type created successfully"),
        ApiResponse(responseCode = "401", description = "Unauthorized access"),
        ApiResponse(responseCode = "400", description = "Invalid request data")
    )
    fun publishBlockType(@RequestBody request: CreateBlockTypeRequest): ResponseEntity<BlockType> {
        val entity = blockTypeService.publishBlockType(request)
        return ResponseEntity.status(HttpStatus.CREATED).body(entity)
    }

    @PutMapping("/{blockTypeId}")
    @Operation(
        summary = "Update an existing block type",
        description = "Updates a block type with the specified ID. Does not allow changing the scope of the block type."
    )
    @ApiResponses(
        ApiResponse(responseCode = "200", description = "Block type updated successfully"),
        ApiResponse(responseCode = "401", description = "Unauthorized access"),
        ApiResponse(responseCode = "404", description = "Block type not found"),
        ApiResponse(responseCode = "400", description = "Invalid request data")
    )
    fun updateBlockType(
        @PathVariable blockTypeId: UUID,
        @RequestBody blockType: BlockType
    ): ResponseEntity<Void> {
        if (blockType.id != blockTypeId) {
            return ResponseEntity.badRequest().build()
        }
        blockTypeService.updateBlockType(blockType)
        return ResponseEntity.ok().build()
    }

    @PutMapping("/{blockTypeId}/archive/{status}")
    @Operation(
        summary = "Archive a block type",
        description = "Archives a block type by its ID. The block type will still be visible to users currently using it but cannot be used in new blocks."
    )
    @ApiResponses(
        ApiResponse(responseCode = "204", description = "Block type archived successfully"),
        ApiResponse(responseCode = "401", description = "Unauthorized access"),
        ApiResponse(responseCode = "404", description = "Block type not found")
    )
    fun updateArchiveStatusByBlockTypeId(
        @PathVariable blockTypeId: UUID,
        @PathVariable status: Boolean
    ): ResponseEntity<Unit> {
        blockTypeService.archiveBlockType(blockTypeId, status)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/key/{key}")
    @Operation(
        summary = "Get block type by key",
        description = "Retrieves a block type by its unique key."
    )
    @ApiResponses(
        ApiResponse(responseCode = "200", description = "Block type retrieved successfully"),
        ApiResponse(responseCode = "401", description = "Unauthorized access"),
        ApiResponse(responseCode = "404", description = "Block type not found")
    )
    fun getBlockTypeByKey(@PathVariable key: String): ResponseEntity<BlockType> {
        val blockTypeEntity = blockTypeService.getEntityByKey(key)
        // Note: Assuming toModel() conversion exists on entity - following the established pattern
        val blockType = blockTypeEntity.toModel()
        return ResponseEntity.ok(blockType)
    }

    @GetMapping("/organisation/{organisationId}")
    @Operation(
        summary = "Get block types for organisation",
        description = "Retrieves all block types associated with a specific organisation."
    )
    @ApiResponses(
        ApiResponse(responseCode = "200", description = "Block types retrieved successfully"),
        ApiResponse(responseCode = "401", description = "Unauthorized access"),
        ApiResponse(responseCode = "404", description = "No block types found for the organisation")
    )
    fun getBlockTypes(
        @PathVariable organisationId: UUID,
    ): ResponseEntity<List<BlockType>> {
        val blockTypes = blockTypeService.getBlockTypes(organisationId)
        return ResponseEntity.ok(blockTypes)
    }

    @PostMapping("/fork")
    @Operation(
        summary = "Fork an existing block type",
        description = "Creates a fork of an existing block type, allowing for modifications."
    )
    @ApiResponses(
        ApiResponse(responseCode = "201", description = "Block type forked successfully"),
        ApiResponse(responseCode = "401", description = "Unauthorized access"),
        ApiResponse(responseCode = "404", description = "Source block type not found"),
        ApiResponse(responseCode = "501", description = "Feature not yet implemented")
    )
    fun forkBlockType(
        @RequestBody source: BlockType,
    ): ResponseEntity<Void> {
        blockTypeService.forkBlockType(source)
        return ResponseEntity.status(HttpStatus.CREATED).build()
    }
}

File: client/components/feature-modules/blocks/controller/block-type.controller.ts
import {
    BlockType,
    CreateBlockTypeRequest,
    GetBlockTypesResponse,
} from "@/components/feature-modules/blocks/interface/block.interface";
import { handleError, validateSession, validateUuid } from "@/lib/util/controller/controller.util";
import { fromError, isResponseError } from "@/lib/util/error/error.util";
import { api } from "@/lib/util/utils";
import { Session } from "@supabase/supabase-js";

/**
 * Publish (create) a block type
 */
export const publishBlockType = async (
    session: Session | null,
    request: CreateBlockTypeRequest
): Promise<BlockType> => {
    try {
        validateSession(session);
        const url = api();

        const response = await fetch(`${url}/v1/block/schema/`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${session.access_token}`,
            },
            body: JSON.stringify(request),
        });

        if (response.ok) return await response.json();
        throw await handleError(
            response,
            (res) => `Failed to publish block type: ${res.status} ${res.statusText}`
        );
    } catch (error) {
        if (isResponseError(error)) throw error;
        throw fromError(error);
    }
};

/**
 * Update an existing block type by id
 */
export const updateBlockType = async (
    session: Session | null,
    blockTypeId: string,
    request: BlockType
): Promise<BlockType> => {
    try {
        validateSession(session);
        validateUuid(blockTypeId);

        const url = api();

        const response = await fetch(`${url}/v1/block/schema/${blockTypeId}`, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${session.access_token}`,
            },
            body: JSON.stringify(request),
        });

        if (response.ok) return await response.json();

        throw await handleError(
            response,
            (res) => `Failed to update block type: ${res.status} ${res.statusText}`
        );
    } catch (error) {
        if (isResponseError(error)) throw error;
        throw fromError(error);
    }
};

/**
 * Update archive status on a block type
 */
export const updateArchiveStatusByBlockTypeId = async (
    session: Session | null,
    blockTypeId: string,
    status: string,
    requestBody: BlockType
): Promise<BlockType> => {
    try {
        validateSession(session);
        validateUuid(blockTypeId);

        const url = api();

        const response = await fetch(
            `${url}/v1/block/schema/${blockTypeId}/archive/${encodeURIComponent(status)}`,
            {
                method: "PUT",
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${session.access_token}`,
                },
                body: JSON.stringify(requestBody),
            }
        );

        if (response.ok) return await response.json();

        throw await handleError(
            response,
            (res) => `Failed to update block type archive status: ${res.status} ${res.statusText}`
        );
    } catch (error) {
        if (isResponseError(error)) throw error;
        throw fromError(error);
    }
};

/**
 * Get block types for an organisation
 */
export const getBlockTypes = async (
    session: Session | null,
    organisationId: string
): Promise<GetBlockTypesResponse> => {
    try {
        validateUuid(organisationId);
        validateSession(session);

        const url = api();

        const response = await fetch(`${url}/v1/block/schema/organisation/${organisationId}`, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${session.access_token}`,
            },
        });

        if (response.ok) return await response.json();

        throw await handleError(
            response,
            (res) => `Failed to fetch block types: ${res.status} ${res.statusText}`
        );
    } catch (error) {
        if (isResponseError(error)) throw error;
        throw fromError(error);
    }
};

/**
 * Get a block type by key
 */
export const getBlockTypeByKey = async (
    session: Session | null,
    key: string
): Promise<BlockType> => {
    try {
        validateSession(session);

        const url = api();

        const response = await fetch(`${url}/v1/block/schema/key/${encodeURIComponent(key)}`, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${session.access_token}`,
            },
        });

        if (response.ok) return await response.json();

        throw await handleError(
            response,
            (res) => `Failed to fetch block type by key: ${res.status} ${res.statusText}`
        );
    } catch (error) {
        if (isResponseError(error)) throw error;
        throw fromError(error);
    }
};

export const lintBlockType = async (session: Session | null, blockType: BlockType) => {
    try {
        validateSession(session);
        const url = api();
        
        const response = await fetch(`${url}/v1/block/schema/lint/`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${session.access_token}`,
            },
            body: JSON.stringify(blockType),
        });

        if (response.ok) return await response.json();

        throw await handleError(
            response,
            (res) => `Failed to lint block type: ${res.status} ${res.statusText}`
        );
    } catch (error) {
        if (isResponseError(error)) throw error;
        throw fromError(error);
    }
};
